// cohere-ai/dist/serialization/resources/finetuning/resources/finetuning/types/Hyperparameters.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Hyperparameters = void 0;
const core = __importStar(require("../../../../../../core"));
const LoraTargetModules_1 = require("./LoraTargetModules");
exports.Hyperparameters = core.serialization.object({
    earlyStoppingPatience: core.serialization.property("early_stopping_patience", core.serialization.number().optional()),
    earlyStoppingThreshold: core.serialization.property("early_stopping_threshold", core.serialization.number().optional()),
    trainBatchSize: core.serialization.property("train_batch_size", core.serialization.number().optional()),
    trainEpochs: core.serialization.property("train_epochs", core.serialization.number().optional()),
    learningRate: core.serialization.property("learning_rate", core.serialization.number().optional()),
    loraAlpha: core.serialization.property("lora_alpha", core.serialization.number().optional()),
    loraRank: core.serialization.property("lora_rank", core.serialization.number().optional()),
    loraTargetModules: core.serialization.property("lora_target_modules", LoraTargetModules_1.LoraTargetModules.optional()),
});
