// cohere-ai/serialization/client/requests/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenizeRequest = exports.SummarizeRequest = exports.RerankRequest = exports.GenerateStreamRequest = exports.GenerateRequest = exports.EmbedRequest = exports.DetokenizeRequest = exports.ClassifyRequest = exports.ChatStreamRequest = exports.ChatRequest = void 0;
var ChatRequest_1 = require("./ChatRequest");
Object.defineProperty(exports, "ChatRequest", { enumerable: true, get: function () { return ChatRequest_1.ChatRequest; } });
var ChatStreamRequest_1 = require("./ChatStreamRequest");
Object.defineProperty(exports, "ChatStreamRequest", { enumerable: true, get: function () { return ChatStreamRequest_1.ChatStreamRequest; } });
var ClassifyRequest_1 = require("./ClassifyRequest");
Object.defineProperty(exports, "ClassifyRequest", { enumerable: true, get: function () { return ClassifyRequest_1.ClassifyRequest; } });
var DetokenizeRequest_1 = require("./DetokenizeRequest");
Object.defineProperty(exports, "DetokenizeRequest", { enumerable: true, get: function () { return DetokenizeRequest_1.DetokenizeRequest; } });
var EmbedRequest_1 = require("./EmbedRequest");
Object.defineProperty(exports, "EmbedRequest", { enumerable: true, get: function () { return EmbedRequest_1.EmbedRequest; } });
var GenerateRequest_1 = require("./GenerateRequest");
Object.defineProperty(exports, "GenerateRequest", { enumerable: true, get: function () { return GenerateRequest_1.GenerateRequest; } });
var GenerateStreamRequest_1 = require("./GenerateStreamRequest");
Object.defineProperty(exports, "GenerateStreamRequest", { enumerable: true, get: function () { return GenerateStreamRequest_1.GenerateStreamRequest; } });
var RerankRequest_1 = require("./RerankRequest");
Object.defineProperty(exports, "RerankRequest", { enumerable: true, get: function () { return RerankRequest_1.RerankRequest; } });
var SummarizeRequest_1 = require("./SummarizeRequest");
Object.defineProperty(exports, "SummarizeRequest", { enumerable: true, get: function () { return SummarizeRequest_1.SummarizeRequest; } });
var TokenizeRequest_1 = require("./TokenizeRequest");
Object.defineProperty(exports, "TokenizeRequest", { enumerable: true, get: function () { return TokenizeRequest_1.TokenizeRequest; } });
