// cohere-ai/serialization/resources/batches/types/Batch.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Batch = void 0;
const core = __importStar(require("../../../../core"));
const BatchStatus_1 = require("./BatchStatus");
exports.Batch = core.serialization.object({
    id: core.serialization.string().optional(),
    name: core.serialization.string(),
    creatorId: core.serialization.property("creator_id", core.serialization.string().optional()),
    orgId: core.serialization.property("org_id", core.serialization.string().optional()),
    status: BatchStatus_1.BatchStatus.optional(),
    createdAt: core.serialization.property("created_at", core.serialization.date().optional()),
    updatedAt: core.serialization.property("updated_at", core.serialization.date().optional()),
    inputDatasetId: core.serialization.property("input_dataset_id", core.serialization.string()),
    outputDatasetId: core.serialization.property("output_dataset_id", core.serialization.string().optional()),
    inputTokens: core.serialization.property("input_tokens", core.serialization.string().optional()),
    outputTokens: core.serialization.property("output_tokens", core.serialization.string().optional()),
    model: core.serialization.string(),
    numRecords: core.serialization.property("num_records", core.serialization.number().optional()),
    numSuccessfulRecords: core.serialization.property("num_successful_records", core.serialization.number().optional()),
    numFailedRecords: core.serialization.property("num_failed_records", core.serialization.number().optional()),
    statusReason: core.serialization.property("status_reason", core.serialization.string().optional()),
});
