// cohere-ai/dist/serialization/types/NonStreamedChatResponse.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.NonStreamedChatResponse = void 0;
const core = __importStar(require("../../core"));
const ApiMeta_1 = require("./ApiMeta");
const ChatCitation_1 = require("./ChatCitation");
const ChatDocument_1 = require("./ChatDocument");
const ChatSearchQuery_1 = require("./ChatSearchQuery");
const ChatSearchResult_1 = require("./ChatSearchResult");
const FinishReason_1 = require("./FinishReason");
const Message_1 = require("./Message");
const ToolCall_1 = require("./ToolCall");
exports.NonStreamedChatResponse = core.serialization.object({
    text: core.serialization.string(),
    generationId: core.serialization.property("generation_id", core.serialization.string().optional()),
    responseId: core.serialization.property("response_id", core.serialization.string().optional()),
    citations: core.serialization.list(ChatCitation_1.ChatCitation).optional(),
    documents: core.serialization.list(ChatDocument_1.ChatDocument).optional(),
    isSearchRequired: core.serialization.property("is_search_required", core.serialization.boolean().optional()),
    searchQueries: core.serialization.property("search_queries", core.serialization.list(ChatSearchQuery_1.ChatSearchQuery).optional()),
    searchResults: core.serialization.property("search_results", core.serialization.list(ChatSearchResult_1.ChatSearchResult).optional()),
    finishReason: core.serialization.property("finish_reason", FinishReason_1.FinishReason.optional()),
    toolCalls: core.serialization.property("tool_calls", core.serialization.list(ToolCall_1.ToolCall).optional()),
    chatHistory: core.serialization.property("chat_history", core.serialization.list(Message_1.Message).optional()),
    meta: ApiMeta_1.ApiMeta.optional(),
});
