// cohere-ai/test/env.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const webpack_1 = __importDefault(require("webpack"));
(0, vitest_1.describe)("test env compatibility", () => {
    (0, vitest_1.test)("webpack", () => __awaiter(void 0, void 0, void 0, function* () {
        yield new Promise((resolve, reject) => {
            (0, webpack_1.default)({
                mode: "production",
                entry: "./src/index.ts",
                module: {
                    rules: [
                        {
                            test: /\.tsx?$/,
                            use: 'ts-loader',
                            exclude: /node_modules/,
                        },
                    ],
                },
                resolve: {
                    extensions: ['.tsx', '.ts', '.js'],
                },
            }, (err, stats) => {
                console.log(stats === null || stats === void 0 ? void 0 : stats.toString());
                try {
                    (0, vitest_1.expect)(err).toBe(null);
                    (0, vitest_1.expect)(stats === null || stats === void 0 ? void 0 : stats.hasErrors()).toBe(false);
                    resolve();
                }
                catch (e) {
                    reject(e);
                }
            });
        });
    }));
});
